### DISCUSSION TITLE
State of reDIP-CIA and reDIP-SID

### DISCUSSION THREAD

**Discussion by jueank** [daglem/reDIP-CIA#3] [General]
Hi @daglem,
me, as just being a simple C64 'user' have discovered your two projects a few months ago. Since there have been no major updates for some time, I wonder what the status of the 2 projects is? Do you consider them as being finished and usable?

What would I have to do to get reDIP-CIA and reDIP-SID up and running?
Do you send (sell) ready built boards to interested people? Or should I order from one of the typical PCB manufacturers with SMD assembly? How do I get the FPGAs programmed or the data onto the flash memory? Your short hints in the readme don't tell me enough what to do. I am inexperienced with FPGAs and the toolchains and additional programming hardware.

**Comment by daglem** [comment]
Hi,

Sorry for the late reply, I've had a vacation.

I am quite happy with how the reDIP SID has turned out, for status see [hardware](https://github.com/daglem/reDIP-SID) and [gateware](https://github.com/daglem/reDIP-SID/tree/master/gateware).

The reDIP CIA hardware has so far not been prototyped, and there is no gateware yet (as you have found out).

I don't think such projects will ever be "finished" - there will always be something one would like to change / improve. For the reDIP SID, I am working on finding ways to further improve the emulation of the MOS 6581 filter. The digital emulation is, to the best of my knowledge, 100% accurate.

So far I have decided against selling boards, and I don't have any spare reDIP SID prototypes. You can order from PCBA manufacturers, however that's going to cost you! Just the BOM cost for a single board is now well north of $20 - the cost of FPGAs and other chips has skyrocketed in the last few years. And you won't get the lowest cost PCB assembly for the board at JLCPCB, PCBWay or similar, due to 0.4mm pitch (and 0201?) components. For the record, I used PCBCart to manufacture prototypes.

As it stands, you'd have to *really* want these boards, as they will be extremely costly compared to MCU based solutions like the [SIDKick pico](https://github.com/frntc/SIDKick-pico).

If you should be ready to part with the several hundred dollars it's probably going to cost to get a few boards assembled, the flash chip can first be programmed with the [Nitro Bootloader](https://github.com/no2fpga/no2bootloader/), after which further programming can be done over USB with [dfu-util](https://dfu-util.sourceforge.net/) (see the [Makefile](https://github.com/daglem/reDIP-SID/blob/master/gateware/Makefile)).

Out of the box, the Nitro Bootloader is [programmed](https://github.com/no2fpga/no2bootloader/tree/master/gateware/ice40-stub) using `make BOARD=redip-sid bootloader`. This depends on  `iceprog`, which in turn depends on an FTDI chip (used on most Lattice eval boards) to program the flash chip over SPI via the reDIP SID programming header. I used an FTDI `C232HM-DDHSL-0` cable, but I think anything with a FT232H or FT2232H chip should work. It is also possible to program the flash chip by other means (it's just SPI, after all), e.g. using OpenOCD on a Raspberry Pi. If I understand correctly, you just have to hold `/RST` low while programming, in order to keep the FPGA off the bus.

**Comment by jueank** [comment]
Hi!
happy new year and thanks a lot for answering! :)
I was pleasantly surprised, that you gave such an in-depth answer!

I'm generally interested in both of your projects. Just because of playing around with new stuff..
In this case, I'm more interested in the ReDIP CIA. As you wrote, there are plenty alternatives for the SID (obviously, because it has the much fancier job and therefore more attention). I have a SIDKick pico at home and an ARMSID. But for the CIA, there's not much (the J-CIA. But that one is boring, because a) it's rather expensive and b) it's closed source / closed hardware. You can only buy it and use it and not build it / tinker with it yourself).

I threw the reDIP CIA into the calculator at JLCPCB, and yes, it's quite expensive to produce with pcb assembly. Partially because of the 201 components (I was already thinking to replace them in KiCad with 402 sized components). Partially because they have strange combinations of allowed pcb color, thickness, number of pcbs when they (jlcpcb) do assembly. Black pcb is quite more expensive than green.

To give you a bit background (in case you want to know) why I'm interested in the reDIP CIA:
I built a 250466+ board; see the attached picture. One of my goals is to use as few original MOS chips as possible. I built a GAL PLA from an open design, as seen on the pictures with 2 nice PLLC28 package chips. I also built a 6510 MOS replacer, so that you can use a 6502 instead of a 6510. That was pretty tough to solder, because of many tiny smd components (201 again). I'm considering, building another 6510 replacement, based on a CPLD. There are 2 designs publicly available. 
The ARMSID of course as a replacement for the SID. Remains the CIAs and the VIC (and the 8701).
So, yeah, I understand that there is just no software/gateware yet for the reDIP CIA. That's a pity...
Do you have concrete plans for the near future in regards to reDIP CIA? Or do you expect someone else to do it? Like with the SID?

![20241224_141533_tn](https://github.com/user-attachments/assets/e364fdc5-4a4b-49da-a7bb-b242f8ff54f6)
![20241224_141601_tn](https://github.com/user-attachments/assets/f995c235-9107-4fe1-bc3f-4958851b923b)

**Reply by daglem** [reply]
It would cost a lot less to manufacture turnkey reDIP CIA boards than reDIP SID boards, but as you found it would still not be super cheap.

Note that in order to replace 0201 components with 0402 on the reDIP CIA, you'd have to use both sides of the board. The goal of the design was to be elegant/minimal, and also to facilitate hot plate reflow for DIYers. The board is 50.80mm x 17.78mm, so the MHP50 hot plate should fit the bill. Seeing that you have experience with assembly, and if you enjoy doing that, you could order bare PCBs (e.g. 3x1 v-score panel) with a stencil. Otherwise, if I get around to ordering a few turnkey prototype boards, I'll let you know.

I will hopefully get to implementing gateware for the reDIP CIA eventually, but don't hold your breath. I don't *expect* anyone else to implement gateware, but open hardware makes it possible for anyone to try. I'm not sure what you mean regarding the reDIP SID. I got invaluable help with auxilliary code for the reDIP SID (bootloader, audio codec configuration) and testing, however the actual SID emulation gateware is entirely my own design.

**Reply by jueank** [reply]
Hi!
I would not mind having some components also on the underside of the board. I'd rather have slightly larger components (402, 603) which are not so hard to work with than these ultra small ones. My limited experience is that more often than not, you have to do manual corrections with the soldering iron after reflowing and then you still have to work with these tiny components. You could do the reflow process for the upper side first and solder some components by hand on the underside afterwards. I just wanted to share my thoughts, no intention to debate it. Everyone has it's own preferences.
Why are you forced to move some components to the underside? Because of the tight placement of the pull-up resistors between the flash mem pads and the THT holes?
reDIP CIA is really the project, which I'm intertested the most in at the moment. I was even thinking what it would mean to dive into CPLD/FPGA programming. Unfortunately, I have zero experience. And this might be a too complex topic for a start. Is there an algorithmic challenge in it? Or is the CIA fully documented  - and how to map it to an FPGA and it's more of a painstaking task that requires concentration? Do you have to tweak timing and fix errors in a long stabilisation process? I have no idea where the challenge lies.....
Btw. sorry for mixing up the work on the gateware for reDIP SID. It was some time ago when I read about it and only remembered that there was another person involved.

![ReDIP CIA](https://github.com/user-attachments/assets/e15d0a95-431b-474c-b52b-8fb8e75700cb)

**Reply by daglem** [reply]
I'm happy with the board in its current form - feel free to make your own redesign. Replacing 0201 components with 0402, I think that due to space constraints you'd have to move most, if not all, of those to the bottom side. BTW the chips next to the pull-up resistors are level shifters, not flash.

I won't go into discussing FPGA implementations here. It's a steep learning curve for sure, you'll just have to start studying.
The MOS 8521 CIA silicon has been documented together with other MOS chip [dissections](http://forum.6502.org/viewtopic.php?f=4&t=7427). This would be a good place to start, together with datasheets, emulators, and test suites.

**Reply by jueank** [reply]
Hey, 
I think there might be a little misunderstanding. I wasn't expecting you to make changes to the original board design. I just wanted to ask what I should pay attention to if I make the changes myself, and then I came to talk about preferences of board designs...
Also, I wanted to mention some aspects of the FPGA implementation of the CIA just in case you want to chat about it. I was not expecting you to teach me about FPGAs in general.

**Comment by mikebgky** [comment]
When I first saw this project I was so impressed that I rushed out and 
bought the components for the build.  I never bothered to check if a PCB 
was available or determine whether I had the skill sets necessary to 
assemble such a project.  The answer is, I do not, but I remain 
impressed.  Hope someone finishes a prototype someday.


------ Original Message ------
From "Dag Lem" ***@***.***>
To "daglem/reDIP-CIA" ***@***.***>
Cc "Subscribed" ***@***.***>
Date 1/16/2025 3:22:21 PM
Subject Re: [daglem/reDIP-CIA] State of reDIP-CIA and reDIP-SID 
(Discussion #3)
